lib/Sema/CSSimplify.cpp

```diff
@@ -3939,9 +3940,19 @@ ConstraintSystem::simplifyKeyPathConstraint(Type keyPathTy,
     break;
   }
   
+  // FIXME: Allow the type to be upcast if the type system has a concrete
+  // KeyPath type assigned to the expression already.
+  if (keyPathBGT) {
+    if (keyPathBGT->getDecl() == getASTContext().getKeyPathDecl())
+      kpDecl = getASTContext().getKeyPathDecl();
+    else if (keyPathBGT->getDecl() == getASTContext().getWritableKeyPathDecl()
+             && capability >= Writable)
+      kpDecl = getASTContext().getWritableKeyPathDecl();
+  }
```

Comment:
What fails without that code?

I tried applying your changes locally and it looked like the tests you're adding all type-checked to WritableKeyPath with or without the change above.